File: 49-group-anagrams/49-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) 
    {
        map<string,int> mpp;
        
        int count=0;
        
        for(auto it : strs)
        {
            string temp = it;
            sort(temp.begin(), temp.end());
            
            if(!mpp.count(temp))
            {
                mpp[temp]=count;
                count++;
            }
        }
        
        vector<vector<string>> ans(count);
        
        for(auto it : strs)
        {
            string temp = it;
            sort(temp.begin(), temp.end());
            
            int b = mpp[temp];
            ans[b].push_back(it);
        }
        
        return ans;
    }
};

File: 49-group-anagrams/49. Group Anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> n = new ArrayList<>();
        HashMap<String, ArrayList<Integer>> mpp = new HashMap<>();

        int i=0;
        for(String it : strs)
        {
            char[] arr = it.toCharArray();
            Arrays.sort(arr);

            String sorted = new String(arr);

            if(!mpp.containsKey(sorted))mpp.put(sorted, new ArrayList<>());
            mpp.get(sorted).add(i);
            i++;
        }

        for(String key : mpp.keySet())
        {
            ArrayList<Integer> curr = mpp.get(key);
            ArrayList<String> temp = new ArrayList<>();

            for(Integer it : curr)
            {
                temp.add(strs[it]);
            }

            n.add(temp);
        }

        return n;
    }
}